File: solutions/2026-06-07/java/2026-06-07-create-binary-tree-from-descriptions.java
/**
 * Problem: Create Binary Tree From Descriptions
 * Link: https://leetcode.com/problems/create-binary-tree-from-descriptions/
 *
 * Summary:
 * This problem asks us to construct a binary tree given a list of parent-child
 * relationships with information about whether the child is left or right.
 * We need to identify the root and build the tree structure.
 *
 * Approach:
 * 1. Identify all unique node values to easily create TreeNode objects as needed.
 * 2. Use two hash maps:
 *    - `nodeMap`: To store TreeNode objects for each value, so we can quickly
 *                 access or create them.
 *    - `parentSet`: To keep track of all values that are children. This helps
 *                   us identify the root node, which is the value that is never
 *                   a child.
 * 3. Iterate through the `descriptions` array:
 *    - For each description `[parentVal, childVal, isLeft]`:
 *      - Get or create the TreeNode for `parentVal` from `nodeMap`.
 *      - Get or create the TreeNode for `childVal` from `nodeMap`.
 *      - If `isLeft` is 1, set the left child of the parent node to the child node.
 *      - If `isLeft` is 0, set the right child of the parent node to the child node.
 *      - Add `childVal` to the `parentSet`.
 * 4. After processing all descriptions, iterate through the keys in `nodeMap`.
 *    The key that is not present in `parentSet` is the root of the tree.
 * 5. Return the TreeNode corresponding to the root value.
 *
 * Time Complexity:
 * O(N), where N is the number of descriptions.
 * We iterate through the descriptions once to build the tree.
 * We iterate through the unique node values (at most 2N) to find the root.
 * Map operations (put, get, containsKey) are O(1) on average.
 *
 * Space Complexity:
 * O(N), where N is the number of unique node values.
 * The `nodeMap` can store up to 2N nodes.
 * The `parentSet` can store up to N child values.
 */

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    /**
     * Constructs a binary tree from a given list of descriptions.
     *
     * @param descriptions A 2D integer array where descriptions[i] = [parenti, childi, isLefti]
     *                     indicates that parenti is the parent of childi, and isLefti determines
     *                     if childi is the left (1) or right (0) child.
     * @return The root node of the constructed binary tree.
     */
    public TreeNode createBinaryTree(int[][] descriptions) {
        // Map to store TreeNode objects for each value. Key: node value, Value: TreeNode object.
        Map<Integer, TreeNode> nodeMap = new HashMap<>();
        // Set to store all values that appear as children. This helps identify the root.
        Set<Integer> childSet = new HashSet<>();

        // Iterate through each description to build the tree structure.
        for (int[] description : descriptions) {
            int parentVal = description[0];
            int childVal = description[1];
            int isLeft = description[2];

            // Get or create the parent TreeNode.
            // The computeIfAbsent method is used here for conciseness:
            // if parentVal is already a key, it returns its value; otherwise, it
            // creates a new TreeNode with parentVal and puts it into the map,
            // then returns the new TreeNode.
            TreeNode parentNode = nodeMap.computeIfAbsent(parentVal, TreeNode::new);

            // Get or create the child TreeNode.
            TreeNode childNode = nodeMap.computeIfAbsent(childVal, TreeNode::new);

            // Set the child as either the left or right child of the parent.
            if (isLeft == 1) {
                parentNode.left = childNode;
            } else {
                parentNode.right = childNode;
            }

            // Add the child's value to the childSet.
            childSet.add(childVal);
        }

        // Find the root node. The root is the node value that is present in nodeMap
        // but NOT present in childSet (i.e., it's never a child).
        TreeNode root = null;
        for (int val : nodeMap.keySet()) {
            if (!childSet.contains(val)) {
                root = nodeMap.get(val);
                break; // Found the root, no need to check further.
            }
        }

        // Return the root of the constructed binary tree.
        return root;
    }
}

File: solutions/2026-06-07/python/2026-06-07-create-binary-tree-from-descriptions.py
```python
# Problem: Create Binary Tree From Descriptions
# Link: https://leetcode.com/problems/create-binary-tree-from-descriptions/
# Approach:
# 1. We need to build a binary tree from a list of parent-child relationships.
# 2. To efficiently access nodes by their values, we'll use a hash map (dictionary in Python) to store TreeNode objects.
# 3. We also need to identify the root node. The root is the node that is a parent but never a child.
#    We can track all child nodes in a set. The node that appears as a parent but not in the set of children is the root.
# 4. Iterate through the `descriptions` list:
#    a. For each `[parent_val, child_val, is_left]`:
#       i. Get or create the `parent_node` and `child_node` from the hash map.
#       ii. If `is_left` is 1, set `parent_node.left = child_node`.
#       iii. If `is_left` is 0, set `parent_node.right = child_node`.
#       iv. Add `child_val` to a set of `children_values` to identify non-root nodes.
# 5. After processing all descriptions, iterate through the `nodes` hash map.
#    The key that is not present in the `children_values` set is the root's value.
# 6. Return the root node.
#
# Time Complexity: O(N), where N is the number of descriptions. We iterate through the descriptions once to build the tree and populate the hash map. Then, we iterate through the hash map (at most N nodes) to find the root.
# Space Complexity: O(N), where N is the number of unique node values. This is due to the hash map storing TreeNode objects and the set storing child values. In the worst case, all values are unique and are either parents or children.

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def createBinaryTree(self, descriptions: list[list[int]]) -> TreeNode | None:
        # A dictionary to store TreeNode objects, keyed by their values for quick access.
        nodes = {}
        # A set to store the values of all child nodes. This helps us identify the root node later.
        children_values = set()

        # Helper function to get or create a TreeNode for a given value.
        def get_or_create_node(val):
            if val not in nodes:
                nodes[val] = TreeNode(val)
            return nodes[val]

        # Iterate through each description to build the tree structure.
        for parent_val, child_val, is_left in descriptions:
            # Get the TreeNode for the parent and child values.
            parent_node = get_or_create_node(parent_val)
            child_node = get_or_create_node(child_val)

            # Add the child's value to the set of children.
            children_values.add(child_val)

            # Assign the child node to its parent based on whether it's a left or right child.
            if is_left:
                parent_node.left = child_node
            else:
                parent_node.right = child_node

        # Find the root node. The root node is the one that exists as a parent but is never a child.
        # Iterate through all the nodes we've created.
        root_node = None
        for val, node in nodes.items():
            # If a node's value is not found in the set of children values, it means it's a root.
            if val not in children_values:
                root_node = node
                # Since the problem guarantees a valid tree, there will be exactly one root.
                break

        # Return the root of the constructed binary tree.
        return root_node

```